packages/puppeteer-core/src/cdp/Input.ts

```diff
@@ -552,57 +554,110 @@ export class CdpMouse extends Mouse {
+    keyboard: CdpKeyboard,
+    public id: number,
+    touchPoint: Protocol.Input.TouchPoint,
+  ) {
     this.#client = client;
     this.#keyboard = keyboard;
+    this.#touchPoint = touchPoint;
   }
 
-  updateClient(client: CDPSession): void {
+  public updateClient(client: CDPSession): void {
     this.#client = client;
   }
 
-  override async touchStart(x: number, y: number): Promise<void> {
+  public async start(): Promise<void> {
+    if (this.#started) {
+      return;
+    }
     await this.#client.send('Input.dispatchTouchEvent', {
       type: 'touchStart',
-      touchPoints: [
-        {
-          x: Math.round(x),
-          y: Math.round(y),
-          radiusX: 0.5,
-          radiusY: 0.5,
-          force: 0.5,
-        },
-      ],
+      touchPoints: [this.#touchPoint],
       modifiers: this.#keyboard._modifiers,
     });
+    this.#started = true;
   }
 
-  override async touchMove(x: number, y: number): Promise<void> {
-    await this.#client.send('Input.dispatchTouchEvent', {
+  public move(x: number, y: number): Promise<void> {
+    this.#touchPoint.x = Math.round(x);
+    this.#touchPoint.y = Math.round(y);
+    return this.#client.send('Input.dispatchTouchEvent', {
       type: 'touchMove',
-      touchPoints: [
-        {
-          x: Math.round(x),
-          y: Math.round(y),
-          radiusX: 0.5,
-          radiusY: 0.5,
-          force: 0.5,
-        },
-      ],
+      touchPoints: [this.#touchPoint],
       modifiers: this.#keyboard._modifiers,
     });
   }
 
-  override async touchEnd(): Promise<void> {
-    await this.#client.send('Input.dispatchTouchEvent', {
+  public end(): Promise<void> {
+    return this.#client.send('Input.dispatchTouchEvent', {
       type: 'touchEnd',
-      touchPoints: [],
+      touchPoints: [this.#touchPoint],
       modifiers: this.#keyboard._modifiers,
     });
   }
 }
+
+/**
+ * @internal
+ */
+export class CdpTouchscreen extends Touchscreen {
+  #client: CDPSession;
+  #keyboard: CdpKeyboard;
+  #touches: CdpTouch[] = [];
+  #idRepository = new TouchPointIdRepository();
+
+  constructor(client: CDPSession, keyboard: CdpKeyboard) {
+    super();
+    this.#client = client;
+    this.#keyboard = keyboard;
+  }
+
+  updateClient(client: CDPSession): void {
+    this.#client = client;
+    this.#touches.forEach(t => {
+      t.updateClient(client);
+    });
+  }
+
+  override async touchStart(x: number, y: number): Promise<Touch> {
+    const id = this.#idRepository.getId();
+    const touchPoint: Protocol.Input.TouchPoint = {
+      x: Math.round(x),
+      y: Math.round(y),
+      radiusX: 0.5,
+      radiusY: 0.5,
+      force: 0.5,
+      id,
+    };
+    const touch = new CdpTouch(this.#client, this.#keyboard, id, touchPoint);
+    await touch.start();
+    this.#touches.push(touch);
+    return touch;
+  }
+
+  override touchMove(x: number, y: number): Promise<void> {
+    const touch = this.#touches[0];
+    if (!touch) {
+      return Promise.resolve();
```

Comment:
@OrKoN I think what I am about to suggest may be breaking, but if no touch is started shouldn't we throw an error?

Reply:
The opposite, actually! Turns out that we currently already throw in this case, as evidenced by [this test](https://github.com/emilefokkema/puppeteer/blob/4ca2a144f5b8f2aa0476ef8ece9348d180c6ca99/test/src/touchscreen.spec.ts#L249). I just changed this so that we still throw in this case.